### PR TITLE
Add Get/SetComObjectData tests

### PR DESCRIPTION
**Pull request by hughbe**
~Relies on https://github.com/dotnet/coreclr/pull/18900~
Fixes dotnet/corefx#31015

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
This work is blocking #31188.

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/ComEventsHelperTests.cs [hunk 1]
```diff
@@ -9,63 +9,63 @@ namespace System.Runtime.InteropServices.Tests
...
         {
-            Assert.Throws<PlatformNotSupportedException>(() => ComEventsHelper.Combine(1, Guid.Empty, 1, null));
+            AssertExtensions.Throws<ArgumentNullException>(null, () => ComEventsHelper.Combine(null, Guid.Empty, 1, null));
         }
 
         [Fact]
         [PlatformSpecific(TestPlatforms.Windows)]
-        [SkipOnTargetFramework(TargetFrameworkMonikers.UapAot, "Throws PlatformNotSupportedException in UapAot")]
-        public void Remove_NullRcwWindows_ThrowsArgumentNullException()
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "ComEventsHelper.Combine is not supported in .NET Core.")]
```
Comment (AaronRobinsonMSFT) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/ComEventsHelperTests.cs]: Sorry, does this say this attribute now say skip on `NetFramework`, but the comment says the function isn't supported on .NET Core?

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/Common/CommonTypes.Windows.cs [hunk 2]
```diff
@@ -2,12 +2,34 @@
...
+#pragma warning disable CS0618 // Type or member is obsolete
+
 namespace System.Runtime.InteropServices.Tests.Common
 {
     [ComImport]
     [Guid("927971f5-0939-11d1-8be1-00c04fd8d503")]
     public interface IComImportObject { }
 
+    [ComImport]
+    [Guid("927971f5-0939-11d1-8be1-00c04fd8d503")]
```
Comment (AaronRobinsonMSFT) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/Common/CommonTypes.Windows.cs]: Guids are all identical. Use `uuidgen /c` from Visual Studio Developer Command prompt to generate guids.

Comment (hughbe) [reply src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/Common/CommonTypes.Windows.cs]: I changed the guids for the interfaces, but doing this with the classes causes test failures:
```
  ----- start 11:43:07.47 ===============  To repro directly: =====================================================
  pushd C:\Users\hughb\Documents\GitHub\corefx\bin\tests\System.Runtime.InteropServices.Tests\netcoreapp-Windows_NT-Deb
  ug-x64\
  call C:\Users\hughb\Documents\GitHub\corefx\bin\testhost\netcoreapp-Windows_NT-Debug-x64\\dotnet.exe xunit.console.ne
  tcore.exe System.Runtime.InteropServices.Tests.dll  -xml testResults.xml -notrait Benchmark=true -notrait category=no
  nnetcoreapptests -notrait category=nonwindowstests  -notrait category=OuterLoop -notrait category=failing
  popd
  ===========================================================================================================
  xUnit.net console test runner (64-bit .NET Core)
  Copyright (C) 2014 Outercurve Foundation.

  Discovering: System.Runtime.InteropServices.Tests
  Discovered:  System.Runtime.InteropServices.Tests
  Starting:    System.Runtime.InteropServices.Tests
     System.Runtime.InteropServices.Tests.IsComObjectTests.IsComObject_Windows_ReturnsExpected [FAIL]
        System.Runtime.InteropServices.COMException : Retrieving the COM class factory for component with CLSID {838CEA
  E6-4C5E-4364-B2F8-B9024B35BC8E} failed due to the following error: 80040154 Class not registered (Exception from HRES
  ULT: 0x80040154 (REGDB_E_CLASSNOTREG)).
        Stack Trace:
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime.InteropServices\tests\System\Runtime\InteropServic
  es\Marshal\IsComObjectTests.Windows.cs(15,0): at System.Runtime.InteropServices.Tests.IsComObjectTests.IsComObject_Wi
  ndows_TestData()+MoveNext()
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Linq\src\System\Linq\Select.cs(135,0): at System.Linq.Enum
  erable.SelectEnumerableIterator`2.MoveNext()
     System.Runtime.InteropServices.Tests.GetComObjectDataTests.GetComObjectData_ValidObject_ReturnsExpected [FAIL]
        System.Runtime.InteropServices.COMException : Retrieving the COM class factory for component with CLSID {838CEA
  E6-4C5E-4364-B2F8-B9024B35BC8E} failed due to the following error: 80040154 Class not registered (Exception from HRES
  ULT: 0x80040154 (REGDB_E_CLASSNOTREG)).
        Stack Trace:
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime.InteropServices\tests\System\Runtime\InteropServic
  es\Marshal\GetComObjectDataTests.Windows.cs(16,0): at System.Runtime.InteropServices.Tests.GetComObjectDataTests.GetC
  omObjectData_ValidObject_ReturnsExpected()
     System.Runtime.InteropServices.Tests.SetComObjectDataTests.SetComObjectData_NullValue_Sucesss [FAIL]
        System.Runtime.InteropServices.COMException : Retrieving the COM class factory for component with CLSID {838CEA
  E6-4C5E-4364-B2F8-B9024B35BC8E} failed due to the following error: 80040154 Class not registered (Exception from HRES
  ULT: 0x80040154 (REGDB_E_CLASSNOTREG)).
        Stack Trace:
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime.InteropServices\tests\System\Runtime\InteropServic
  es\Marshal\SetComObjectDataTests.Windows.cs(32,0): at System.Runtime.InteropServices.Tests.SetComObjectDataTests.SetC
  omObjectData_NullValue_Sucesss()
     System.Runtime.InteropServices.Tests.SetComObjectDataTests.SetComObjectData_NonNullValue_Sucesss [FAIL]
        System.Runtime.InteropServices.COMException : Retrieving the COM class factory for component with CLSID {838CEA
  E6-4C5E-4364-B2F8-B9024B35BC8E} failed due to the following error: 80040154 Class not registered (Exception from HRES
  ULT: 0x80040154 (REGDB_E_CLASSNOTREG)).
        Stack Trace:
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime.InteropServices\tests\System\Runtime\InteropServic
  es\Marshal\SetComObjectDataTests.Windows.cs(16,0): at System.Runtime.InteropServices.Tests.SetComObjectDataTests.SetC
  omObjectData_NonNullValue_Sucesss()
     System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventHandler_TargetNotIConnectionIConnectionPointCo
  ntainer_ThrowsInvalidCastException [FAIL]
        System.Runtime.InteropServices.COMException : Retrieving the COM class factory for component with CLSID {838CEA
  E6-4C5E-4364-B2F8-B9024B35BC8E} failed due to the following error: 80040154 Class not registered (Exception from HRES
  ULT: 0x80040154 (REGDB_E_CLASSNOTREG)).
        Stack Trace:
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime.InteropServices\tests\System\Runtime\InteropServic
  es\ComAwareEventInfoTests.Windows.cs(112,0): at System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventH
  andler_TargetNotIConnectionIConnectionPointContainer_ThrowsInvalidCastException()
     System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventHandler_ComObjectWithMultipleComEventInterface
  Attribute_ThrowsAmbiguousMatchException [FAIL]
        System.Runtime.InteropServices.COMException : Retrieving the COM class factory for component with CLSID {838CEA
  E6-4C5E-4364-B2F8-B9024B35BC8E} failed due to the following error: 80040154 Class not registered (Exception from HRES
  ULT: 0x80040154 (REGDB_E_CLASSNOTREG)).
        Stack Trace:
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime.InteropServices\tests\System\Runtime\InteropServic
  es\ComAwareEventInfoTests.Windows.cs(57,0): at System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventHa
  ndler_ComObjectWithMultipleComEventInterfaceAttribute_ThrowsAmbiguousMatchException()
     System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventHandler_NoSuchSourceTypeEventInterface_ThrowsA
  rgumentNullException [FAIL]
        System.Runtime.InteropServices.COMException : Retrieving the COM class factory for component with CLSID {838CEA
  E6-4C5E-4364-B2F8-B9024B35BC8E} failed due to the following error: 80040154 Class not registered (Exception from HRES
  ULT: 0x80040154 (REGDB_E_CLASSNOTREG)).
        Stack Trace:
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime.InteropServices\tests\System\Runtime\InteropServic
  es\ComAwareEventInfoTests.Windows.cs(84,0): at System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventHa
  ndler_NoSuchSourceTypeEventInterface_ThrowsArgumentNullException()
     System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventHandler_NoDispIdAttribute_ThrowsInvalidOperati
  onException [FAIL]
        System.Runtime.InteropServices.COMException : Retrieving the COM class factory for component with CLSID {838CEA
  E6-4C5E-4364-B2F8-B9024B35BC8E} failed due to the following error: 80040154 Class not registered (Exception from HRES
  ULT: 0x80040154 (REGDB_E_CLASSNOTREG)).
        Stack Trace:
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime.InteropServices\tests\System\Runtime\InteropServic
  es\ComAwareEventInfoTests.Windows.cs(101,0): at System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventH
  andler_NoDispIdAttribute_ThrowsInvalidOperationException()
     System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventHandler_ComObjectWithoutComEventInterfaceAttri
  bute_ThrowsInvalidOperationException [FAIL]
        System.Runtime.InteropServices.COMException : Retrieving the COM class factory for component with CLSID {838CEA
  E6-4C5E-4364-B2F8-B9024B35BC8E} failed due to the following error: 80040154 Class not registered (Exception from HRES
  ULT: 0x80040154 (REGDB_E_CLASSNOTREG)).
        Stack Trace:
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime.InteropServices\tests\System\Runtime\InteropServic
  es\ComAwareEventInfoTests.Windows.cs(32,0): at System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventHa
  ndler_ComObjectWithoutComEventInterfaceAttribute_ThrowsInvalidOperationException()
     System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventHandler_NullSourceTypeEventInterface_ThrowsNul
  lReferenceException [FAIL]
        System.Runtime.InteropServices.COMException : Retrieving the COM class factory for component with CLSID {838CEA
  E6-4C5E-4364-B2F8-B9024B35BC8E} failed due to the following error: 80040154 Class not registered (Exception from HRES
  ULT: 0x80040154 (REGDB_E_CLASSNOTREG)).
        Stack Trace:
           C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime.InteropServices\tests\System\Runtime\InteropServic
  es\ComAwareEventInfoTests.Windows.cs(67,0): at System.Runtime.InteropServices.Tests.ComAwareEventInfoTests.AddEventHa
  ndler_NullSourceTypeEventInterface_ThrowsNullReferenceException()
  Finished:    System.Runtime.InteropServices.Tests
```

**Comment by danmosemsft** [comment]
`MSBUILD : error MSB4166: Child node "2" exited prematurely. Shutting down. Diagnostic information may be found in files in "C:\Users\dotnet-bot\AppData\Local\Temp\" and will be named MSBuild_*.failure.txt. This location can be changed by setting the MSBUILDDEBUGPATH environment variable to a different directory.`

@dotnet-bot test Windows x64 Debug Build please

**Comment by danmosemsft** [comment]
@hughbe could you please rebase

**Comment by hughbe** [comment]
Fixed merge conflict... wondering why the the csproj format can't be similar to normal projects that pick up new files automatically :'(

File: src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/ComEventsHelperTests.cs [hunk 3]
```diff
@@ -9,63 +9,63 @@ namespace System.Runtime.InteropServices.Tests
...
+        [PlatformSpecific(TestPlatforms.Windows)]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.UapAot, "Throws PlatformNotSupportedException in UapAot")]
+        public void Remove_NullRcw_ThrowsArgumentNullException()
+        {
+            AssertExtensions.Throws<ArgumentNullException>(null, () => ComEventsHelper.Remove(null, Guid.Empty, 1, null));
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
         [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "ComEventsHelper.Combine is not supported in .NET Core.")]
```
Comment (luqunl) [inline src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/ComEventsHelperTests.cs]: Do we need to skip this on .NETCore?

**Comment by karelz** [comment]
@luqunl is there any reason why you didn't merge it?

**Comment by AaronRobinsonMSFT** [comment]
@karelz This actually should have been merged from what I recall.

**Comment by karelz** [comment]
@AaronRobinsonMSFT @luqunl can one of you please merge it then?

**Comment by karelz** [comment]
@luqunl yes, once there is 1 MS engineer code review (excluding author), the PR can be merged. It is preferable for area owners to merge PRs - they know best how to edit commit comments.

**Comment by luqunl** [comment]
Thanks @karelz . Got it.

**Pull request merged**